### PR TITLE
sopt: new package

### PR DESCRIPTION
**Pull request by tkoskela**
Add a package for the sparse optimisation library [sopt](https://github.com/astro-informatics/sopt/releases/tag/v4.2.0). It is used for astrophysics imaging problems. I will add an application in a separate PR #46839 .

### CODE REVIEW

**Review by 20DM — APPROVED** [review]
Thanks - looks good to me!

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]
@tkoskela @20DM: When enabling the benchmarks variant using `+benchmarks`, cmake fails:
```py
  >> 12    CMake Error at CMakeLists.txt:107 (include):
     13      include could not find requested file:
     14
     15        AddBenchmark
     16
     17
  >> 18    CMake Error at cpp/benchmarks/CMakeLists.txt:1 (add_benchmark):
     19      Unknown CMake command "add_benchmark".
```

File: var/spack/repos/builtin/packages/sopt/package.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    variant("onnxrt", default=False, description="Build with Tensorflow support using onnx")
+
+    depends_on("cmake@3")
+    depends_on("eigen@3.4")
+    depends_on("libtiff@4.5")
+    depends_on("mpi", when="+mpi")
+    depends_on("catch2@3.4", when="+tests")
+    depends_on("benchmark@1.8", when="+benchmarks")
+    depends_on("onnx@1.16", when="+onnxrt")
+    depends_on("doxygen@1.9", when="+docs")
```
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/sopt/package.py]: I have a problem building this specific version of `doxygen` with `gcc-13`:
```py
/home/bkaindl/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/spdlog-1.14.1-go7oyje5evdqx44xoiqvxbwbl4mijqy7/include/spdlog/fmt/fmt.h:24:14: fatal error: spdlog/fmt/bundled/core.h: No such file or directory
   24 |     #include <spdlog/fmt/bundled/core.h>
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

One can't build using gcc-13 with this version of doxygen, it seems.

- Therefore, see it as a problem to depend on this specific version of doxygen
  - Doxygen likely just misses a `conficts("%gcc@13:", when="@:1.9")` (or so, depends on specifics)

- Do you really need this specific version of doxygen (and specifically, only this version)?
  - This also applies to all the dependencies above
  - Usually, you should never depend on a single specific version, usually you depend on a range, that is usually open-ended: at least version x.x, or up to version x.x.

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]

**Comment by tkoskela** [comment]
> @tkoskela @20DM: When enabling the benchmarks variant using `+benchmarks`, cmake fails:
> 
> ```python
>   >> 12    CMake Error at CMakeLists.txt:107 (include):
>      13      include could not find requested file:
>      14
>      15        AddBenchmark
>      16
>      17
>   >> 18    CMake Error at cpp/benchmarks/CMakeLists.txt:1 (add_benchmark):
>      19      Unknown CMake command "add_benchmark".
> ```

Thanks for catching this! I've added a patch to fix it

File: var/spack/repos/builtin/packages/sopt/package.py (hunk 1 above)
Comment (tkoskela) [reply var/spack/repos/builtin/packages/sopt/package.py]: It looks like we can build the documentation just fine with doxygen@1.12 (that doesn't give me any issues with gcc-13), so happy to remove that specific version. Is it better to have "at least version 1.12" in the spec, or leave it unspecified?

Comment (tkoskela) [reply var/spack/repos/builtin/packages/sopt/package.py]: I've added version ranges where it seemed reasonable to do so.

**Review by bernhardkaindl — COMMENTED** [review]
> Thanks for catching this! I've added a patch to fix it

@tkoskela It looks like the CI error at https://github.com/spack/spack/actions/runs/11219405757/job/31185310501?pr=46837 wants you to append `?full_index=1` to the URL of the patch:
```py
 PKG-DIRECTIVES: 1 issue found
1. patch URL in package sopt must end with ?full_index=1
    https://github.com/astro-informatics/sopt/commit/836171f32d39a3fbc1147d6c302a08a61f842fee.patch 
```

**Review by bernhardkaindl — DISMISSED** [review]
Build results on Ubuntu 22.04.4 LTS:
```py
bin/spack find --variants sopt
-- linux-ubuntu22.04-skylake / clang@14.0.0 ---------------------
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp+tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi+onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo+mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs+examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage+docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks+coverage~docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0+benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db

-- linux-ubuntu22.04-skylake / clang@15.0.7 ---------------------
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp+tests build_system=cmake build_type=Release generator=make
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi+onnxrt~openmp~tests build_system=cmake build_type=Release generator=make
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo+mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make
sopt@4.2.0~benchmarks~coverage~docs+examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make
sopt@4.2.0~benchmarks~coverage+docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks+coverage~docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make
sopt@4.2.0+benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db

-- linux-ubuntu22.04-skylake / gcc@8.4.0 ------------------------
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp+tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi+onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs+examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage+docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks+coverage~docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0+benchmarks~coverage~docs+examples~ipo+mpi~onnxrt+openmp+tests build_system=cmake build_type=Release generator=make patches=00729db

-- linux-ubuntu22.04-skylake / gcc@9.5.0 ------------------------
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp+tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi+onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs+examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage+docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks+coverage~docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0+benchmarks~coverage~docs+examples~ipo+mpi~onnxrt+openmp+tests build_system=cmake build_type=Release generator=make patches=00729db

-- linux-ubuntu22.04-skylake / gcc@10.5.0 -----------------------
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp+tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi+onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs+examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage+docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks+coverage~docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0+benchmarks~coverage~docs+examples~ipo+mpi~onnxrt+openmp+tests build_system=cmake build_type=Release generator=make patches=00729db

-- linux-ubuntu22.04-skylake / gcc@11.4.0 -----------------------
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp+tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi+onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs+examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0+benchmarks~coverage~docs+examples~ipo+mpi~onnxrt+openmp+tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0+benchmarks+coverage+docs+examples~ipo+mpi~onnxrt+openmp+tests build_system=cmake build_type=Release generator=make patches=00729db

-- linux-ubuntu22.04-skylake / gcc@12.3.0 -----------------------
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp+tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi+onnxrt~openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs~examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs+examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0+benchmarks~coverage~docs+examples~ipo+mpi~onnxrt+openmp+tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0+benchmarks+coverage+docs+examples~ipo+mpi~onnxrt+openmp+tests build_system=cmake build_type=Release generator=make patches=00729db

-- linux-ubuntu22.04-skylake / gcc@13.1.0 -----------------------
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt~openmp+tests build_system=cmake build_type=Release generator=make
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo~mpi+onnxrt~openmp~tests build_system=cmake build_type=Release generator=make
sopt@4.2.0~benchmarks~coverage~docs~examples~ipo+mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make
sopt@4.2.0~benchmarks~coverage~docs~examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks~coverage~docs+examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make
sopt@4.2.0~benchmarks~coverage~docs+examples+ipo~mpi~onnxrt+openmp~tests build_system=cmake build_type=Release generator=make patches=00729db
sopt@4.2.0~benchmarks+coverage~docs~examples~ipo~mpi~onnxrt~openmp~tests build_system=cmake build_type=Release generator=make
sopt@4.2.0+benchmarks+coverage+docs+examples~ipo+mpi~onnxrt+openmp+tests build_system=cmake build_type=Release generator=make patches=00729db
```
Generated by https://github.com/spack/build-quality-tools/blob/main/build_pr_changes.py
```py
build-quality-tools/build_pr_changes.py -b sopt -may
```

**Review by bernhardkaindl — APPROVED** [review]
@tkoskela When updating the URL, the data itself of course also changed.

**Pull request merged**